File: swift/Sources/FilenameParse.swift
/*
    djent - A reimplementation of Fourmilab's ent with several improvements.
    
    Copyright (C) 2017  David Johnston
    Translated to Swift 2026

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    Contact. David Johnston dj@deadhat.com
*/

import Foundation

/// Results from parsing a filename for test conditions
struct FilenameParsed {
    var voltage: Double = 0.0
    var temperature: Double = 0.0
    var deviceID: String = ""
    var process: String = ""
}

/// Look for the _<int>p<int>V_ voltage pattern in a string.
/// Returns the matched substring or nil.
func findVoltagePattern(in str: String) -> String? {
    let chars = Array(str)
    let len = chars.count
    var start = 0
    var state = 1
    var pos = 0

    while pos < len {
        let c = chars[pos]
        switch state {
        case 1: // looking for _
            if c == "_" { state = 2; start = pos }
            pos += 1
        case 2: // first digit
            if c.isNumber { state = 3 } else { state = 1 }
            pos += 1
        case 3: // rest of first int
            if c.isNumber { /* stay */ }
            else if c == "p" || c == "." { state = 4 }
            else { state = 1 }
            pos += 1
        case 4: // first digit after decimal
            if c.isNumber { state = 5 } else { state = 1 }
            pos += 1
        case 5: // rest of second int
            if c.isNumber { /* stay */ }
            else if c == "V" { state = 6 }
            else { state = 1 }
            pos += 1
        case 6: // trailing _ or .
            if c == "_" || c == "." {
                let end = pos
                return String(chars[start...end])
            } else { state = 1 }
            pos += 1
        default:
            pos += 1
        }
    }
    return nil
}

/// Look for the _<int>p<int>C_ temperature pattern in a string.
func findTemperaturePattern(in str: String) -> String? {
    let chars = Array(str)
    let len = chars.count
    var start = 0
    var state = 1
    var pos = 0

    while pos < len {
        let c = chars[pos]
        switch state {
        case 1:
            if c == "_" { state = 2; start = pos }
            pos += 1
        case 2: // first digit or minus sign
            if c.isNumber || c == "-" { state = 3 } else { state = 1 }
            pos += 1
        case 3: // rest of first int
            if c.isNumber { /* stay */ }
            else if c == "p" || c == "." { state = 4 }
            else { state = 1 }
            pos += 1
        case 4:
            if c.isNumber { state = 5 } else { state = 1 }
            pos += 1
        case 5:
            if c.isNumber { /* stay */ }
            else if c == "C" { state = 6 }
            else { state = 1 }
            pos += 1
        case 6:
            if c == "_" || c == "." {
                let end = pos
                return String(chars[start...end])
            } else { state = 1 }
            pos += 1
        default:
            pos += 1
        }
    }
    return nil
}

/// Look for the _CID-<ID>_ pattern in a string.
func findCIDPattern(in str: String) -> String? {
    let chars = Array(str)
    let len = chars.count
    var start = 0
    var state = 1
    var pos = 0

    while pos < len {
        let c = chars[pos]
        switch state {
        case 1:
            if c == "_" { state = 2; start = pos }
            pos += 1
        case 2:
            state = c == "C" ? 3 : 1; pos += 1
        case 3:
            state = c == "I" ? 4 : 1; pos += 1
        case 4:
            state = c == "D" ? 5 : 1; pos += 1
        case 5:
            state = c == "-" ? 6 : 1; pos += 1
        case 6: // first char of ID
            if c != "_" { state = 7 } else { state = 1 }
            pos += 1
        case 7: // rest of ID
            if c == "_" {
                let end = pos
                return String(chars[start...end])
            }
            pos += 1
        default:
            pos += 1
        }
    }
    return nil
}

/// Look for the _PROC-<name>_ pattern in a string.
func findProcPattern(in str: String) -> String? {
    let chars = Array(str)
    let len = chars.count
    var start = 0
    var state = 1
    var pos = 0

    while pos < len {
        let c = chars[pos]
        switch state {
        case 1:
            if c == "_" { state = 2; start = pos }
            pos += 1
        case 2:
            state = c == "P" ? 3 : 1; pos += 1
        case 3:
            state = c == "R" ? 4 : 1; pos += 1
        case 4:
            state = c == "O" ? 5 : 1; pos += 1
        case 5:
            state = c == "C" ? 6 : 1; pos += 1
        case 6:
            state = c == "-" ? 7 : 1; pos += 1
        case 7:
            if c != "_" { state = 8 } else { state = 1 }
            pos += 1
        case 8:
            if c == "_" {
                let end = pos
                return String(chars[start...end])
            }
            pos += 1
        default:
            pos += 1
        }
    }
    return nil
}

/// Parse a filename for voltage, temperature, CID and process fields.
func parseFilename(_ filename: String) -> FilenameParsed {
    var result = FilenameParsed()

    if var match = findVoltagePattern(in: filename) {
        match = match.replacingOccurrences(of: "p", with: ".")
        // Remove leading _ and trailing V_ or V.
        let trimmed = match.trimmingCharacters(in: CharacterSet(charactersIn: "_."))
        if let vStr = trimmed.components(separatedBy: "V").first,
           let v = Double(vStr) {
            result.voltage = v
        }
    } else {
        fputs("Regex error scanning for _<num>p<num>V_:\n", stderr)
    }

    if var match = findTemperaturePattern(in: filename) {
        match = match.replacingOccurrences(of: "p", with: ".")
        let trimmed = match.trimmingCharacters(in: CharacterSet(charactersIn: "_."))
        if let tStr = trimmed.components(separatedBy: "C").first,
           let t = Double(tStr) {
            result.temperature = t
        }
    } else {
        fputs("Regex error scanning for _<num>p<num>C_:\n", stderr)
    }

    if let match = findCIDPattern(in: filename) {
        // Extract ID after _CID- and before trailing _
        let inner = match.dropFirst(5).dropLast(1) // drop "_CID-" and trailing "_"
        result.deviceID = String(inner)
    } else {
        fputs("Regex error scanning for _CID-<ID>_:\n", stderr)
    }

    if let match = findProcPattern(in: filename) {
        let inner = match.dropFirst(6).dropLast(1) // drop "_PROC-" and trailing "_"
        result.process = String(inner)
    } else {
        fputs("Regex error scanning for _PROC-<ID>_:\n", stderr)
    }

    return result
}

File: zig/src/filename_parse.zig
//! djent - Filename parsing for test condition extraction.
//!
//! Copyright (C) 2017  David Johnston
//! Translated to Zig 0.15.x, 2025
//!
//! Contact. David Johnston dj@deadhat.com

const std = @import("std");

pub const FilenameParsed = struct {
    voltage: f64 = 0.0,
    temperature: f64 = 0.0,
    device_id: [256]u8 = [_]u8{0} ** 256,
    device_id_len: usize = 0,
    process: [256]u8 = [_]u8{0} ** 256,
    process_len: usize = 0,
};

const MatchResult = struct { start: usize, end: usize };

/// Look for _<int>p<int>V_ voltage pattern. Returns matched range or null.
pub fn findVPattern(str: []const u8) ?MatchResult {
    var state: u8 = 1;
    var pos: usize = 0;
    var start: usize = 0;

    while (pos < str.len) {
        const c = str[pos];
        switch (state) {
            1 => {
                if (c == '_') { state = 2; start = pos; }
                pos += 1;
            },
            2 => {
                state = if (std.ascii.isDigit(c)) 3 else 1;
                pos += 1;
            },
            3 => {
                if (std.ascii.isDigit(c)) {
                    // stay
                } else if (c == 'p' or c == '.') {
                    state = 4;
                } else { state = 1; }
                pos += 1;
            },
            4 => {
                state = if (std.ascii.isDigit(c)) 5 else 1;
                pos += 1;
            },
            5 => {
                if (std.ascii.isDigit(c)) {
                    // stay
                } else if (c == 'V') { state = 6; } else { state = 1; }
                pos += 1;
            },
            6 => {
                if (c == '_' or c == '.') return .{ .start = start, .end = pos };
                state = 1;
                pos += 1;
            },
            else => { pos += 1; },
        }
    }
    return null;
}

/// Look for _<int>p<int>C_ temperature pattern.
pub fn findTPattern(str: []const u8) ?MatchResult {
    var state: u8 = 1;
    var pos: usize = 0;
    var start: usize = 0;

    while (pos < str.len) {
        const c = str[pos];
        switch (state) {
            1 => {
                if (c == '_') { state = 2; start = pos; }
                pos += 1;
            },
            2 => {
                state = if (std.ascii.isDigit(c) or c == '-') 3 else 1;
                pos += 1;
            },
            3 => {
                if (std.ascii.isDigit(c)) {} else if (c == 'p' or c == '.') { state = 4; } else { state = 1; }
                pos += 1;
            },
            4 => {
                state = if (std.ascii.isDigit(c)) 5 else 1;
                pos += 1;
            },
            5 => {
                if (std.ascii.isDigit(c)) {} else if (c == 'C') { state = 6; } else { state = 1; }
                pos += 1;
            },
            6 => {
                if (c == '_' or c == '.') return .{ .start = start, .end = pos };
                state = 1;
                pos += 1;
            },
            else => { pos += 1; },
        }
    }
    return null;
}

/// Look for _CID-<ID>_ pattern.
pub fn findCIDPattern(str: []const u8) ?MatchResult {
    var state: u8 = 1;
    var pos: usize = 0;
    var start: usize = 0;

    while (pos < str.len) {
        const c = str[pos];
        switch (state) {
            1 => { if (c == '_') { state = 2; start = pos; } pos += 1; },
            2 => { state = if (c == 'C') 3 else 1; pos += 1; },
            3 => { state = if (c == 'I') 4 else 1; pos += 1; },
            4 => { state = if (c == 'D') 5 else 1; pos += 1; },
            5 => { state = if (c == '-') 6 else 1; pos += 1; },
            6 => { state = if (c != '_') 7 else 1; pos += 1; },
            7 => {
                if (c == '_') return .{ .start = start, .end = pos };
                pos += 1;
            },
            else => { pos += 1; },
        }
    }
    return null;
}

/// Look for _PROC-<n>_ pattern.
pub fn findProcPattern(str: []const u8) ?MatchResult {
    var state: u8 = 1;
    var pos: usize = 0;
    var start: usize = 0;

    while (pos < str.len) {
        const c = str[pos];
        switch (state) {
            1 => { if (c == '_') { state = 2; start = pos; } pos += 1; },
            2 => { state = if (c == 'P') 3 else 1; pos += 1; },
            3 => { state = if (c == 'R') 4 else 1; pos += 1; },
            4 => { state = if (c == 'O') 5 else 1; pos += 1; },
            5 => { state = if (c == 'C') 6 else 1; pos += 1; },
            6 => { state = if (c == '-') 7 else 1; pos += 1; },
            7 => { state = if (c != '_') 8 else 1; pos += 1; },
            8 => {
                if (c == '_') return .{ .start = start, .end = pos };
                pos += 1;
            },
            else => { pos += 1; },
        }
    }
    return null;
}

/// Parse a float from a matched pattern like "_1p2V_" -> 1.2
fn parseMatchedFloat(str: []const u8, suffix: u8) f64 {
    var buf: [256]u8 = undefined;
    var len: usize = 0;
    var started = false;

    for (str) |c| {
        if (c == '_' or c == '.') {
            if (started) break;
            continue;
        }
        if (c == suffix) break;
        started = true;
        buf[len] = if (c == 'p') '.' else c;
        len += 1;
    }

    return std.fmt.parseFloat(f64, buf[0..len]) catch 0.0;
}

/// Parse a filename for voltage, temperature, CID and process fields.
pub fn parseFilename(filename: []const u8) FilenameParsed {
    var result = FilenameParsed{};

    if (findVPattern(filename)) |m| {
        result.voltage = parseMatchedFloat(filename[m.start .. m.end + 1], 'V');
    } else {
        std.debug.print("Regex error scanning for _<num>p<num>V_:\n", .{});
    }

    if (findTPattern(filename)) |m| {
        result.temperature = parseMatchedFloat(filename[m.start .. m.end + 1], 'C');
    } else {
        std.debug.print("Regex error scanning for _<num>p<num>C_:\n", .{});
    }

    if (findCIDPattern(filename)) |m| {
        const id_start = m.start + 5; // skip "_CID-"
        const id_end = m.end;
        const id_len = id_end - id_start;
        if (id_len > 0 and id_len <= 255) {
            @memcpy(result.device_id[0..id_len], filename[id_start..id_end]);
            result.device_id_len = id_len;
        }
    } else {
        std.debug.print("Regex error scanning for _CID-<ID>_:\n", .{});
    }

    if (findProcPattern(filename)) |m| {
        const id_start = m.start + 6; // skip "_PROC-"
        const id_end = m.end;
        const id_len = id_end - id_start;
        if (id_len > 0 and id_len <= 255) {
            @memcpy(result.process[0..id_len], filename[id_start..id_end]);
            result.process_len = id_len;
        }
    } else {
        std.debug.print("Regex error scanning for _PROC-<ID>_:\n", .{});
    }

    return result;
}
